nvbench/detail/measure_cupti.cu

```diff
@@ -0,0 +1,288 @@
+  };
+};
+
+template <>
+struct metric_info<metric_ids::global_store_efficiency>
+{
+  static constexpr const char *metric_name =
+    "smsp__sass_average_data_bytes_per_sector_mem_global_op_st.pct";
+
+  static constexpr const char *summary = "Global Store Efficiency (HW)";
+  static constexpr const char *hint = "percentage";
+  static constexpr const char *short_name = "StoreEff";
+
+  static constexpr const char *description =
+    "Ratio of requested global memory store throughput to required global "
+    "memory store throughput expressed as percentage.";
+
+  static constexpr double divider = 100.0;
+
+  static bool is_collected(nvbench::state &m_state)
+  {
+    return m_state.is_stores_efficiency_collected();
+  };
+};
+
+template <>
+struct metric_info<metric_ids::l1_hit_rate>
+{
+  static constexpr const char *metric_name = "l1tex__t_sector_hit_rate.pct";
+  static constexpr const char *summary     = "L1 Cache Hit Rate (HW)";
+  static constexpr const char *hint        = "percentage";
+  static constexpr const char *short_name  = "L1HitRate";
+  static constexpr const char *description = "Hit rate at L1 cache.";
+  static constexpr double divider          = 100.0;
+
+  static bool is_collected(nvbench::state &m_state)
+  {
+    return m_state.is_l1_hit_rate_collected();
+  };
+};
+
+template <>
+struct metric_info<metric_ids::l2_hit_rate>
+{
+  static constexpr const char *metric_name = "lts__t_sector_hit_rate.pct";
+  static constexpr const char *summary     = "L2 Cache Hit Rate (HW)";
+  static constexpr const char *hint        = "percentage";
+  static constexpr const char *short_name  = "L2HitRate";
+  static constexpr const char *description = "Hit rate at L2 cache.";
+  static constexpr double divider          = 100.0;
+
+  static bool is_collected(nvbench::state &m_state)
+  {
+    return m_state.is_l2_hit_rate_collected();
+  };
+};
+
+
+template <metric_ids id>
+void add_metric(nvbench::state &state,
+                std::vector<std::string> &metrics)
+{
+  if (metric_info<id>::is_collected(state))
+  {
+    metrics.emplace_back(metric_info<id>::metric_name);
+  }
+
+  add_metric<static_cast<metric_ids>(id + 1)>(state, metrics);
+}
+
+
+template <>
+void add_metric<metric_ids::count>(nvbench::state &m_state,
+                                   std::vector<std::string> &metrics)
+{}
+
+
+std::vector<std::string> add_metrics(nvbench::state &state)
+{
+  std::vector<std::string> metrics;
+  metrics.reserve(metric_ids::count);
+
+  add_metric<metric_ids::dram_peak_sustained_throughput>(state, metrics);
+  return metrics;
+}
+
+
+measure_cupti_base::measure_cupti_base(state &exec_state)
+  : m_state{exec_state}
+{
+  try
+  {
+    if (const std::optional<nvbench::device_info> &device = m_state.get_device();
+      device)
+    {
+      m_cupti = cupti_profiler(device, add_metrics(m_state));
+
+      if (!m_cupti->initialized())
+      {
+        m_cupti.reset();
+      }
+    }
+  }
+  catch (...)
+  {
+    m_cupti.reset();
```

Comment:
Should this be logged?

Reply:
For reference, here's how to print a warning to the log: https://github.com/NVIDIA/nvbench/blob/main/nvbench/detail/measure_cold.cu#L158-L196